homeassistant/helpers/discovery.py

```diff
@@ -87,20 +89,51 @@ def load_platform(hass, component, platform, discovered=None,
 
     Use `listen_platform` to register a callback for these events.
     """
-    def discover_platform():
-        """Discover platform job."""
+    fire_coroutine_threadsafe(
+        async_load_platform(hass, component, platform,
+                            discovered, hass_config), hass.loop)
+
+
+@asyncio.coroutine
+def async_load_platform(hass, component, platform, discovered=None,
+                        hass_config=None):
+    """Load a component and platform dynamically.
+
+    Target components will be loaded and an EVENT_PLATFORM_DISCOVERED will be
+    fired to load the platform. The event will contain:
+        { ATTR_SERVICE = LOAD_PLATFORM + '.' + <<component>>
+          ATTR_PLATFORM = <<platform>>
+          ATTR_DISCOVERED = <<discovery info>> }
+
+    Use `listen_platform` to register a callback for these events.
+
+    Warning: Do not yield from this inside a setup method to avoid a dead lock.
+    Use `hass.loop.create_task(async_load_platform(..))` instead.
+
+    This method is a coroutine.
+    """
+    did_lock = False
+    if hasattr(hass, 'setup_lock') and hass.setup_lock.locked():
+        did_lock = True
+        yield from hass.setup_lock.acquire()
```

Comment:
This is very nice, I have two of the hub components where I had to wait for `HOME_ASSISTANT_START`, which could probably now be removed